packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts

```diff
@@ -340,30 +340,20 @@ export class OdspDocumentDeltaConnection extends DocumentDeltaConnection impleme
     /**
      * Disconnect from the websocket
      */
-    public disconnect(socketProtocolError: boolean = false) {
-        // We set the closed flag as a part of the contract for overriding the disconnect method.  This is used by
-        // DocumentDeltaConnection to determine if emitting on the socket is allowed, which is important since
-        // OdspDocumentDeltaConnection reuses the socket rather than truly disconnecting it.  Note that below we may
-        // still send disconnect_document which is allowed; this is only intended to prevent normal messages from
-        // being emitted.
-        this.closed = true;
-
-        if (this.socketReferenceKey !== undefined) {
-            const key = this.socketReferenceKey;
-            this.socketReferenceKey = undefined;
-
-            const reason = "client closing connection";
-
-            if (this.enableMultiplexing && !socketProtocolError && this.hasDetails) {
-                // tell the server we are disconnecting this client from the document
-                this.socket.emit("disconnect_document", this.clientId, this.documentId);
-            }
+    protected disconnectCore(socketProtocolError: boolean, reason: string) {
+        const key = this.socketReferenceKey;
+        assert(key !== undefined, "reetrancy not supported!");
+        this.socketReferenceKey = undefined;
+
+        if (!socketProtocolError && this.hasDetails) {
+            // tell the server we are disconnecting this client from the document
+            this.socket.emit("disconnect_document", this.clientId, this.documentId);
+        }
 
-            OdspDocumentDeltaConnection.removeSocketIoReference(key, socketProtocolError, reason);
+        OdspDocumentDeltaConnection.removeSocketIoReference(key, socketProtocolError, reason);
 
-            // RemoveSocketIoReference() above raises "disconnect" event on socket for socketProtocolError === true
-            // If it's not critical error, we want to raise event on this object only.
-            this.emit("disconnect", reason);
-        }
+        // RemoveSocketIoReference() above raises "disconnect" event on socket for socketProtocolError === true
+        // If it's not critical error, we want to raise event on this object only.
+        this.emit("disconnect", reason);
```

Comment:
I think the call in `disconnect()` to `removeTrackedListeners()` will remove `"disconnect"` listeners before this has a chance to emit?

Reply:
Nevermind, I get it now - was confusing the socket object and the document delta connection object here.

Reply:
The issue has description of these 3 paths:
1. If the call comes from close(), then yes, it removes all listeners - we do not want reentrancy here - this connection is being silently closed. 
2. If the call comes from early handlers, then we do raise an event, and then remove all listeners. This path should not happen (i.e. there should be no listeners), but somehow (based on telemetry) we do have listeners. Propagating reason will help us understand from where call happens.
3. "error" handler is special, as we do call disconnect in it, but by this time client should have closed connection. If it did not, it works like # 2 above - raises event and closes connection.